src/main/java/trade/wayruha/mexc/ws/WebSocketClient.java

```diff
@@ -72,77 +76,103 @@ public class WebSocketClient<T> extends WebSocketListener {
+            log.error("{} Failed to send message: {}", logPrefix, request);
             closeOnError(null);
         }
     }
 
     @SneakyThrows
     public void subscribe(Collection<String> newChannels) {
-        if (this.channels.size() + newChannels.size() > WEB_SOCKET_MAX_CHANNELS_PER_CONNECTION)
-            throw new IllegalArgumentException("Can't subscribe to more then " + WEB_SOCKET_MAX_CHANNELS_PER_CONNECTION + "channels in single WS connection");
+        if (this.channels.size() + newChannels.size() > this.config.getWebSocketMaxReconnectAttemptNumber())
+            throw new IllegalArgumentException("Can't subscribe to more then " + WEB_SOCKET_MAX_CHANNELS_PER_CONNECTION + " channels in single WS connection");
         final Message msg = new Message(Action.SUBSCRIPTION, newChannels);
         final String content = objectMapper.writeValueAsString(msg);
-        this.send(content);
+        this.sendRequest(content);
         this.channels.addAll(newChannels);
     }
 
     @SneakyThrows
     public void ping(Collection<String> channels) {
         final Message msg = new Message(Action.PING, channels);
         final String content = objectMapper.writeValueAsString(msg);
-        this.send(content);
-        log.debug("{} ping was send to channels {} ", logPrefix, channels);
+        this.sendRequest(content);
+        log.trace("{} Ping was send to channels {} ", logPrefix, channels);
     }
 
     @SneakyThrows
     public void unsubscribe(Collection<String> channels) {
         final Message msg = new Message(Action.UNSUBSCRIPTION, channels);
         final String content = objectMapper.writeValueAsString(msg);
-        send(content);
+        sendRequest(content);
         this.channels.removeAll(channels);
     }
 
     public void close() {
-        log.debug("{} closing", logPrefix);
+        log.debug("{} Closing WS channels: {}", logPrefix, this.channels);
+        state = WSState.IDLE;
         if (webSocket != null) {
             webSocket.cancel();
             webSocket = null;
         }
-        state = WSState.IDLE;
         this.channels.clear();
         scheduledPingTask.cancel(false);
     }
 
-    public void reConnect() {
-        this.close();
-        this.connect(this.channels);
+    public boolean reConnect() {
+        var success = false;
+        this.canceledChannels = new HashSet<>();
+        this.canceledChannels.addAll(this.channels);
+        while (!success &&
+                (this.config.isWebSocketReconnectAlways() || currentReconnectAttemptIndex.incrementAndGet() < this.config.getWebSocketMaxReconnectAttemptNumber())) {
+            try {
+                log.debug("{} Try to reconnect to WebSocket channels: {}. Attempt #{}",
+                        logPrefix, this.canceledChannels, currentReconnectAttemptIndex.get());
+                this.close();
+                this.connect(this.canceledChannels);
+                success = true;
+                this.canceledChannels.clear();
```

Comment:
if we move this statement down outside of while-loop (e.g., to line 168) then it would be possible to make this.cancelledChannels method-local variable instead of a class field. The only other usage is line 220 (onFailure()) will be redundant if we do it. 
The lesser fields we have - the less complex and fragile the class is. So, please think about it. If you agree - let's do it.

Reply:
No, it won't work due to async way for evoking method reConnect(). More often it evoked in onFailure() method that is consumer for WS events

I alredy refactored all this stuff after 2hours of debugging and checking all kind of cases.